runtime/onert/core/src/compiler/train/StaticGradientShapeInferer.cc

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "StaticGradientShapeInferer.h"
+#include "util/ShapeInference.h"
+#include "util/logging.h"
+
+#include <misc/polymorphic_downcast.h>
+
+#include <sstream>
+#include <stdexcept>
+
+namespace onert
+{
+namespace compiler
+{
+namespace train
+{
+
+void StaticGradientShapeInferer::infer()
+{
+  // It is not determined to iterate in reverse order.
+  auto sorted_ops = _lowered_subg->graph().topolSortOperations();
+  for (auto it = sorted_ops.rbegin(); it != sorted_ops.rend(); ++it)
+  {
+    const auto op_idx = *it;
+    const auto &op = _lowered_subg->trainable_graph().operation(op_idx);
```

Comment:
@ragmani I guess `trainable_graph()` is the only change with existing StaticShapeInferer. Am I right?

Reply:
I didn't understand what you said.

Reply:
- I guess `StaticGradientShapeInferer.cc` is almost identical to `StaticShapeInferer.cc`.
- Most code will be just a copy of `StaticGradientShapeInferer`.
- I guess the only change would be `StaticGradientShapeInferer` iterates `trainable_graph()` in reverse order.

Am I right?

Reply:
No, shapes of gradient tensors are different from tensors for inference. Most code will be different from `StaticShapeInferer`.

Reply:
> No, shapes of gradient tensors are different from tensors for inference. Most code will be different from `StaticShapeInferer`.

Thank you. I would like to get this information. I will see the difference from forthcoming PRs.